From b0d960f102c998d8231c0ee48952b488f10864ac Mon Sep 17 00:00:00 2001
From: Adrien Ferrand <adferrand@users.noreply.github.com>
Date: Wed, 1 May 2019 00:37:23 +0200
Subject: [PATCH] Send a POST-as-GET request to query registration in ACME v2
 (#6993)

* Send a post-as-get request to query registration

* Add comments. Add again a line.

* Prepare code for future PR about post-as-get
---
diff --git a/acme/client.py b/acme/client.py
index a41787756f..5a8fd88ae9 100644
--- a/acme/client.py
+++ b/acme/client.py
@@ -123,15 +123,6 @@ def deactivate_registration(self, regr):
         """
         return self.update_registration(regr, update={'status': 'deactivated'})

-    def query_registration(self, regr):
-        """Query server about registration.
-
-        :param messages.RegistrationResource: Existing Registration
-            Resource.
-
-        """
-        return self._send_recv_regr(regr, messages.UpdateRegistration())
-
     def _authzr_from_response(self, response, identifier=None, uri=None):
         authzr = messages.AuthorizationResource(
             body=messages.Authorization.from_json(response.json()),
@@ -276,6 +267,15 @@ def register(self, new_reg=None):
         # pylint: disable=no-member
         return self._regr_from_response(response)

+    def query_registration(self, regr):
+        """Query server about registration.
+
+        :param messages.RegistrationResource: Existing Registration
+            Resource.
+
+        """
+        return self._send_recv_regr(regr, messages.UpdateRegistration())
+
     def agree_to_tos(self, regr):
         """Agree to the terms-of-service.

@@ -603,10 +603,13 @@ def query_registration(self, regr):
             Resource.

         """
-        self.net.account = regr
-        updated_regr = super(ClientV2, self).query_registration(regr)
-        self.net.account = updated_regr
-        return updated_regr
+        self.net.account = regr  # See certbot/certbot#6258
+        # ACME v2 requires to use a POST-as-GET request (POST an empty JWS) here.
+        # This is done by passing None instead of an empty UpdateRegistration to _post().
+        response = self._post(regr.uri, None)
+        self.net.account = self._regr_from_response(response, uri=regr.uri,
+                                                    terms_of_service=regr.terms_of_service)
+        return self.net.account

     def update_registration(self, regr, update=None):
         """Update registration.
